liquidctl/driver/kraken_three_x.py

```diff
@@ -0,0 +1,327 @@
+    'backwards-moving-alternating-5':       (0x05, 0x05, 4, 1, 2),
+    'backwards-moving-alternating-6':       (0x05, 0x06, 4, 1, 2),
+    'pulse':                                (0x06, 0x00, 5, 1, 8),
+    'breathing':                            (0x07, 0x00, 6, 1, 8),
+    'super-breathing':                      (0x03, 0x00,10, 1, 8),
+    'candle':                               (0x08, 0x00, 0, 1, 1),
+    'starry-night':                         (0x09, 0x00, 5, 1, 1),
+    'rainbow-flow':                         (0x0b, 0x00, 2, 0, 0),
+    'super-rainbow':                        (0x0c, 0x00, 2, 0, 0),
+    'rainbow-pulse':                        (0x0d, 0x00, 2, 0, 0),
+    'backwards-rainbow-flow':               (0x0b, 0x00, 2, 0, 0),
+    'backwards-super-rainbow':              (0x0c, 0x00, 2, 0, 0),
+    'backwards-rainbow-pulse':              (0x0b, 0x00, 2, 0, 0),
+    'loading':                              (0x10, 0x00, 8, 1, 1),
+    'tai-chi':                              (0x0e, 0x00, 7, 1, 2),
+    'water-cooler':                         (0x0f, 0x00, 6, 2, 2),
+    'wings':                                (None, 0x00,11, 1, 1),
+}
+_STATIC_VALUE = {
+    0x02: 0x08,
+    0x04: 0x01
+}
+_SPEED_VALUE = {
+    # (slowest, slow, normal, fast, fastest)
+    0: ([0x32, 0x00], [0x32, 0x00], [0x32, 0x00], [0x32, 0x00], [0x32, 0x00]),
+    1: ([0x50, 0x00], [0x3c, 0x00], [0x28, 0x00], [0x14, 0x00], [0x0a, 0x00]),
+    2: ([0x5e, 0x01], [0x2c, 0x01], [0xfa, 0x00], [0x96, 0x00], [0x50, 0x00]),
+    3: ([0x40, 0x06], [0x14, 0x05], [0xe8, 0x03], [0x20, 0x03], [0x58, 0x02]),
+    4: ([0x20, 0x03], [0xbc, 0x02], [0xf4, 0x01], [0x90, 0x01], [0x2c, 0x01]),
+    5: ([0x19, 0x00], [0x14, 0x00], [0x0f, 0x00], [0x07, 0x00], [0x04, 0x00]),
+    6: ([0x28, 0x00], [0x1e, 0x00], [0x14, 0x00], [0x0a, 0x00], [0x04, 0x00]),
+    7: ([0x32, 0x00], [0x28, 0x00], [0x1e, 0x00], [0x14, 0x00], [0x0a, 0x00]),
+    8: ([0x14, 0x00], [0x14, 0x00], [0x14, 0x00], [0x14, 0x00], [0x14, 0x00]),
+    9: ([0x00, 0x00], [0x00, 0x00], [0x00, 0x00], [0x00, 0x00], [0x00, 0x00]),
+    10:([0x37, 0x00], [0x28, 0x00], [0x19, 0x00], [0x0a, 0x00], [0x00, 0x00]),
+    11:([0x6e, 0x00], [0x53, 0x00], [0x39, 0x00], [0x2e, 0x00], [0x20, 0x00]),
+}
+_ANIMATION_SPEEDS = {
+    'slowest': 0x0,
+    'slower': 0x1,
+    'normal': 0x2,
+    'faster': 0x3,
+    'fastest': 0x4,
+}
+_ACCESSORY_NAMES = {
+    0x01: "HUE+ LED Strip",
+    0x02: "AER RGB 1",
+    0x04: "HUE 2 LED Strip 300 mm",
+    0x05: "HUE 2 LED Strip 250 mm",
+    0x06: "HUE 2 LED Strip 200 mm",
+    0x08: "HUE 2 Cable Comb",
+    0x0a: "HUE 2 Underglow 200 mm",
+    0x0b: "AER RGB 2 120 mm",
+    0x0c: "AER RGB 2 140 mm",
+    0x10: "Kraken X3 Pump Ring",
+    0x11: "Kraken X3 Pump Logo",
+}
+
+
+class KrakenThreeXDriver(UsbHidDriver):
+    """liquidctl driver for Kraken X3 devices from NZXT."""
+
+    SUPPORTED_DEVICES = [
+        (0x1e71, 0x2007, None, 'NZXT Kraken X3 Pump (X53, X63 or X73) (experimental)', {})
```

Comment:
I took this name from the cam device description but I will change it later.

As someone who only knows the `X[4-7]2`, `X[5-7]3` and `Z[6-7]3` devices, a generation naming convention is more confusing rather than naming the devices with their official product names (or suffixes). Especially with the new Z devices. 

I think the `X[5-7]3` notation you mentioned in your comments is the best option. 
And with this naming notation it would be more easy to navigate through the source files:

| old | new |
| :--- | :--- |
|`nzxt-kraken-x-3rd-generation.md` | `nzxt-kraken-x2-devices.md` |
|`-` | `nzxt-kraken-x3-devices.md` |
| `-` | `nzxt-kraken-z3-devices.md` |
|`kraken_two.py` | `kraken_two_x.py` |
|`-` | `kraken_three_x.py` |
|`-` | `kraken_three_z.py` |

Or maybe even name the driver files like the doc files or vice versa. 

Applying a global naming convention for all kraken files would be a great improvement. 
But since there is no global convention just yet, I would suggest to leave this as it is ("Kraken X3 devices") and change all kraken files at once after merging this pull request.

Reply:
My reasoning in the past for going with `NZXT Kraken X (<possible models here>)` was that most users only have a single device, so generation isn't as important as device type (some "Kraken X").  The specific model (X53 vs. X73) would be useful for those with two coolers, but that information is not available anyway.

While I agree that `X[4-7]2` is pretty clear for us programmers, "regular" users would be pretty confused.  And it would be redundant to keep it and and explicit list of possible models.

So, I would prefer to stick with `NZXT Kraken X (X53, X63 or X73)` and (looking forward) `NZXT Kraken Z (Z63 or Z73)`.

---

Regarding the module and driver names, that's something I thought about changing, but decided against to avoid breaking programs already written to the current (or even older) liquidctl APIs.

In a related note, I think that the Z[6-7]3 driver will belong to the same module (and probably documentation page as well) as the X[5-7]3 one, since they appear to share the same base protocol.  The same happens in kraken_two.py, where we have both X[4-7]2 and M22 drivers.

But I agree that using the generation in documentation titles can be a bit confusing, as it seemly doesn't match the model names.

Anyway, here's what I propose:

 - `kraken_three.py`:
    + `KrakenX3Driver(UsbHidDriver)` and `KrakenZ3Driver(KrakenX3Driver)`
    + (already a small improvement over `KrakenTwoDriver`)
 - (option 1) `fourth-generation-krakens.md` and `third-generation-krakens.md`:
 - (or option 2) `nzxt-kraken-3.md` and `nzxt-kraken-2.md`:
    + regardless, each documenting the entire family (X/Z and X/M, respectively)
 - `kraken_two.py` left as is.

I honestly like option 1 best because is more readable and actually meaning something.  And confusion should be minimal given that most will reach those pages from the index, where each specific model is listed explicitly.  The documentation page itself also explains the generation and list the devices that belong to it.

Reply:
Sorry I completely agree that `NZXT Kraken X (<possible models here>)` is the best option for users in any output! I thought this would be clear after my first sentence (; Everything afterwards referred to the source code and comments.

> In a related note, I think that the Z[6-7]3 driver will belong to the same module (and probably documentation page as well) as the X[5-7]3 one, since they appear to share the same base protocol. The same happens in kraken_two.py, where we have both X[4-7]2 and M22 drivers.

Personally I didn't like the structure of the smart module since I needed to scroll so much to understand what's going on, I would prefer one shared driver file for both and separate files for x and z devices. This way you keep the files relatively short and clean. Especially with the probably way more complicated Z code.
Maybe this would be possible with one `kraken_three.py` file and two `kraken_three_x/z.py` files?

If option 2 does not actually mean something, why name the driver file name `kraken_three.py`? At first glance I did not like option 2 as well but it makes sense if you name the drivers this way. And as you mentioned the vast majority of people will access those file via the readme file. If you like to name it after the generation, then I would suggest to name the driver this way as well. 

However, this is your project and just my opinion (: I would suggest to keep it as it is for now and you change the names as you like after merging and before releasing a new version?

Reply:
> Personally I didn't like the structure of the smart module since I needed to scroll so much to understand what's going on

The Smart Device module is indeed quite long.  But I think it mostly has to do with having two distinct generations in a single module.  
_(note to self: that might still be fixable, as it's a newer module and, so far, I'm not aware of any external users of that module)_

But I still think that Kraken X[5-7]3 and Z[6-7]3 belong in a single module.  The only expected differences with the Z[6-7]3 are the additional 'fan' speed channel (minimal change), a `NotImplemented` override for `set_color` and a yet to be specified new API for sending GIFs.

If it weren't for the new API for GIFs both families would even belong in the same driver.

> If option 2 does not actually mean something, why name the driver file name kraken_three.py?

Only to follow the now set-in-stone `kraken_two` module.  But that was a mistake – I didn't know about the Kraken X[46]0 back then – so let's not follow that.  I can't change it without breaking some stuff, but maybe we can do better here.

How about `kraken4.py` or `kraken_gen4.py` for the module name, and `forth-generation-krakens.md` for the documentation?  
_(I think I like `kraken4.py` better for a module name)._

I guess we could also (internally and from now on only) claim to "index from zero" and make it `kraken3.py`, meaning both [XZ][5-7]3 and forth-generation.  But that standard could then be confusing if another family (for example `seasonic.py`) of devices reached a second generation using a "V2" marketing suffix: in this example we would end up with `seasonic1.py` for a "NZXT E850 V2" power supply.

> However, this is your project and just my opinion (: I would suggest to keep it as it is for now and you change the names as you like after merging and before releasing a new version?

Sure, let's leave changing the name for last (though probably just _before_ the merge).

And thanks for taking the time to discuss this with me.  I think consistent and understandable names are important, and feedback helps a lot with this.